librz/core/cmd/cmd_search_rop.c

```diff
@@ -727,3 +730,389 @@ static void rop_classify(RzCore *core, Sdb *db, RzList /*<char *>*/ *ropList, co
+	if (list) {
+		list->free = &rz_rop_constraint_free;
+	}
+	return list;
+}
+
+static void skip_whitespace(const char *str, int *idx) {
+	while (str[*idx] == ' ' || str[*idx] == '\t' || str[*idx] == '\n' || str[*idx] == '\r') {
+		(*idx)++;
+	}
+}
+
+static bool parse_eof(const char *str, int idx) {
+	skip_whitespace(str, &idx);
+	return str[idx] == '\0';
+}
+
+static bool parse_il_equal(char *str, int *idx) {
+	skip_whitespace(str, idx);
+	if (*idx >= strlen(str))
+		return false;
+	if (str[*idx] == '=') {
+		(*idx)++;
+		return true;
+	}
+	return false;
+}
+
+static char *parse_register(char *str, int *idx) {
+	char reg[256] = { 0 };
+	int reg_idx = 0;
+
+	skip_whitespace(str, idx);
+
+	while (isalnum(str[*idx]) || str[*idx] == '_') {
+		reg[reg_idx++] = str[*idx];
+		(*idx)++;
+	}
+
+	// Check if the register is correct for the given architecture.
+	if (reg_idx == 0) {
+		return NULL;
+	}
+
+	return strdup(reg);
+}
+
+static bool parse_constant(const char *str, int *idx, unsigned long long *value) {
+	int base = 10;
+	int neg = 0;
+	char num_str[256] = { 0 };
+	int num_idx = 0;
+
+	skip_whitespace(str, idx);
+
+	if (str[*idx] == '-') {
+		neg = 1;
+		(*idx)++;
+	}
+
+	skip_whitespace(str, idx);
+
+	if (str[*idx] == '0' && (str[*idx + 1] == 'x' || str[*idx + 1] == 'X')) {
+		base = 16;
+		*idx += 2;
+	}
+
+	while (isdigit(str[*idx]) || (base == 16 && isxdigit(str[*idx]))) {
+		num_str[num_idx++] = str[*idx];
+		(*idx)++;
+	}
+
+	if (num_idx == 0) {
+		return false;
+	}
+
+	*value = strtoull(num_str, NULL, base);
+	if (neg) {
+		*value = -*value;
+	}
+
+	return true;
+}
+
+static bool parse_reg_to_const(RzCore *core, char *str, RzRopConstraint *rop_constraint) {
+	int idx = 0;
+	char *dst_reg = parse_register(str, &idx);
+	if (!dst_reg) {
+		return false;
+	}
+
+	if (!parse_il_equal(str, &idx)) {
+		free(dst_reg);
+		return false;
+	}
+
+	unsigned long long const_value;
+	if (!parse_constant(str, &idx, &const_value)) {
+		free(dst_reg);
+		return false;
+	}
+
+	if (!parse_eof(str, idx)) {
+		free(dst_reg);
+		return false;
+	}
+
+	rop_constraint->type = MOV_CONST;
+	rop_constraint->args[DST_REG] = dst_reg;
+	rop_constraint->args[SRC_REG] = NULL;
+	char value_str[256];
+	sprintf(value_str, "%llu", const_value);
+	rop_constraint->args[DST_CONST] = strdup(value_str);
+	return true;
+}
+
+static bool parse_reg_to_reg(RzCore *core, char *str, RzRopConstraint *rop_constraint) {
+	int idx = 0;
+	char *dst_reg = parse_register(str, &idx);
+	if (!dst_reg) {
+		return false;
+	}
+
+	if (!parse_il_equal(str, &idx)) {
+		free(dst_reg);
+		return false;
+	}
+
+	char *src_reg = parse_register(str, &idx);
+	if (!src_reg) {
+		free(dst_reg);
+		return false;
+	}
+
+	if (!parse_eof(str, idx)) {
+		free(dst_reg);
+		return false;
+	}
+
+	rop_constraint->type = MOV_REG;
+	rop_constraint->args[DST_REG] = dst_reg;
+	rop_constraint->args[SRC_REG] = src_reg;
+	return true;
+}
+
+static bool parse_il_op(RzList *args, const char *str, int *idx) {
```

Comment:
This kind of code is better to be implemented in a general way and moved into `librz/il`, also covered with tests. @wargio @thestr4ng3r @DMaroo @Rot127  any thoughts?

Reply:
Or into `librz/core/cil.c` where printing is already implemented.

Reply:
I have already started writing unit tc for these APIs

Reply:
I don't get why we are parsing the IL as a string. Why can't we just pass an `RzILOpEffect *` for every instruction instead of parsing the stringified IL.

Reply:
This is not completely parsed to IL yet. I have just stringified `RzILInstructionType`, the plan is to store the registers and op as `RzILOpEffect` as you have mentioned.

Reply:
No, I would like if there wasn't any parsing at all. Rizin is a single application and it should be able to pass around data without needing to stringify it and then (re)parse it. I must be missing something, because I don't get why any sort of parsing is necessary.

Reply:
I agree with @XVilka. Better move this logic to `librz/il`.

@DMaroo The ROP gadget constraints are given as string by the user. And need to be translated to their RzIL representation for matching. @giridharprasath This was the design you had in mind, correct?

Reply:
Yes, that's right 👍